cobalt/migrate_storage_record/migrate.cc

```diff
@@ -0,0 +1,398 @@
+          record->Delete();
+        }
+      },
+      record.get()));
+  auto bytes = std::vector<uint8_t>(record->GetSize());
+  int read_result =
+      record->Read(reinterpret_cast<char*>(bytes.data()), bytes.size());
+  if (read_result != bytes.size()) {
+    LOG(WARNING) << "Storage record size mismatch " << read_result << " "
+                 << bytes.size();
+    return nullptr;
+  }
+
+  std::string version(reinterpret_cast<const char*>(bytes.data()), 4);
+  if (version != "SAV1") {
+    return nullptr;
+  }
+
+  auto storage_data = std::make_unique<cobalt::storage::Storage>();
+  if (!storage_data->ParseFromArray(
+          reinterpret_cast<const char*>(bytes.data() + 4), bytes.size() - 4)) {
+    LOG(ERROR) << "Unable to parse storage with size " << bytes.size();
+    return nullptr;
+  }
+  return storage_data;
+}
+
+void SetCanonicalCookieCallback(std::atomic_int* pending,
+                                base::OnceClosure* all_done_callback,
+                                net::CookieAccessResult set_cookie_result) {
+  --(*pending);
+  if (pending->load() == 0) {
+    delete pending;
+    std::move(*all_done_callback).Run();
+    delete all_done_callback;
+  }
+}
+
+void MigrateCookies(base::OnceClosure callback,
+                    content::RenderFrameHost* render_frame_host,
+                    std::unique_ptr<cobalt::storage::Storage> storage_data) {
+  if (!storage_data) {
+    std::move(callback).Run();
+    return;
+  }
+
+  content::StoragePartition* storage_partition =
+      render_frame_host->GetStoragePartition();
+  DCHECK(storage_partition);
+  if (!storage_partition) {
+    std::move(callback).Run();
+    return;
+  }
+  network::mojom::CookieManager* cookie_manager =
+      storage_partition->GetCookieManagerForBrowserProcess();
+  DCHECK(cookie_manager);
+  if (!cookie_manager) {
+    std::move(callback).Run();
+    return;
+  }
+
+  LOG(INFO) << "Migrating cookies.";
+  std::atomic_int* pending = new std::atomic_int();
+  pending->store(storage_data->cookies_size());
+  base::OnceClosure* all_done_callback =
+      new base::OnceClosure(std::move(callback));
+  for (auto& c : storage_data->cookies()) {
+    GURL url("https://www" + c.domain() + c.path());
+    auto cookie = net::CanonicalCookie::FromStorage(
+        c.name(), c.value(), c.domain(), c.path(),
+        base::Time::FromInternalValue(c.creation_time_us()),
+        base::Time::FromInternalValue(c.expiration_time_us()),
+        base::Time::FromInternalValue(c.last_access_time_us()),
+        base::Time::FromInternalValue(c.creation_time_us()), c.secure(),
+        c.http_only(), net::CookieSameSite::NO_RESTRICTION,
+        net::COOKIE_PRIORITY_DEFAULT, false,
+        absl::optional<net::CookiePartitionKey>(),
+        net::CookieSourceScheme::kUnset, url::PORT_UNSPECIFIED);
+    cookie_manager->SetCanonicalCookie(
+        *cookie, url, net::CookieOptions::MakeAllInclusive(),
+        base::BindOnce(&SetCanonicalCookieCallback, pending,
```

Comment:
I'm not sure I understand the part of passing |pending| (a naked pointer) to a BindOnce callback -- didn't the compiler complain ? (Telling you to use base::Unretained or similar?)

In any case this pattern is dangerous, I'd suggest creating a `std::unique_ptr<std::atomic_int>` and `std::move()` it into this callback, which would make the ownership clear, and then you can remove the `delete` in SetCanonicalCookieCallback().

When you do that, I'd suggest inlining SetCanonicalCookieCallback() here like you do elsewhere -- I think is clearer to use a lambda if the body is small enough.